feat(docs_tutorials): Token Portal & Uniswap Tutorial (#2726)
AztecProtocol/aztec-packages: docs/docs/dev_docs/tutorials/token_portal/withdrawing_from_l1.md

---

> title: Withdrawing from L1

Reviewer: Should this be "Withdrawing from L2?" or "Withdrawing to L1?"

---

This is where we have tokens on Aztec and want to withdraw them back to L1 (i.e. burn them on L2 and mint on L1). Withdrawing from L1 will be public.

## Withdrawing publicly

Go back to your `main.nr` and paste this:

#include_code exit_to_l1_public /yarn-project/noir-contracts/src/contracts/token_bridge_contract/src/main.nr rust

> For this to work we will need this helper function:

Reviewer 2: In
```suggestion
For this to work we will need this helper function, in `util.nr`:
```

#include_code get_withdraw_content_hash /yarn-project/noir-contracts/src/contracts/token_bridge_contract/src/util.nr rust

**What’s happening here?**

> The `exit_to_l1_public` function enables anyone to withdraw their L2 tokens back to L1 publicly. This is done by burning tokens on L2 and then creating an l2->l1 message.

Reviewer: ```suggestion
The `exit_to_l1_public` function enables anyone to withdraw their L2 tokens back to L1 publicly. This is done by burning tokens on L2 and then creating an L2->L1 message.
```

> 1. Like with our deposit function, we need to create the l2 to l1 message. The content is the _amount_ to burn, the recipient address, and who can execute the withdraw on the L1 portal on behalf of the user. It can be `0x0` for anyone, or a specified address.

Reviewer: ```suggestion
1. Like with our deposit function, we need to create the L2 to L1 message. The content is the _amount_ to burn, the recipient address, and who can execute the withdraw on the L1 portal on behalf of the user. It can be `0x0` for anyone, or a specified address.
```

> 2. `context.message_portal()` passes this content to the kernel circuit which creates the proof for the transaction. The kernel circuit then adds the sender (the L2 address of the bridge + version of aztec) and the recipient (the portal to the L2 address + the chain ID of L1) under the hood, to create the message which gets added as rollup calldata by the sequencer and is stored in the outbox for consumption.

Reviewer: ```suggestion
2. `context.message_portal()` passes this content to the [kernel circuit](../../../concepts/advanced/circuits/kernels/public_kernel.md) which creates the proof for the transaction. The kernel circuit then adds the sender (the L2 address of the bridge + version of aztec) and the recipient (the portal to the L2 address + the chain ID of L1) under the hood, to create the message which gets added as rollup calldata by the sequencer and is stored in the outbox for consumption.
```
I am assuming this is the public kernel circuit

> 3. Finally, you also burn the tokens on L2! Note that it burning is done at the end to follow the check effects interaction pattern. Note that the caller has to first approve the bridge contract to burn tokens on its behalf. Refer to [burn_public function on the token contract](https://github.com/AztecProtocol/dev-rel/blob/main/tutorials/token-contract/README.md#burn_public). The nonce parameter refers to the approval message that the user creates - also refer to [authorizing token spends here](https://github.com/AztecProtocol/dev-rel/blob/main/tutorials/token-contract/README.md#authorizing-token-spends).

Reviewer 2: copy-pasted links that were left over. please remove

Author: dont we want these?

Reviewer: or update to internal docs links

Reviewer 2: actually yea I second Josh -> update to the internal docs!

Because public functions are executed by the sequencer while private methods are executed locally, all public calls are always done _after_ all private calls are completed. So first the burn would happen and only later the sequencer asserts that the token is same. The sequencer just sees a request to `execute_assert_token_is_same` and therefore has no context on what the appropriate private method was. If the assertion fails, then the kernel circuit will fail to create a proof and hence the transaction will be dropped.

Once again, a user must sign an approval message to let the contract burn tokens on their behalf. The nonce refers to this approval message.

For both the public and private flow, we use the same mechanism to determine the content hash. This is because on L1, things are public anyway. The only different between the two functions is that in the private domain we have to nullify user’s notes where as in the public domain we subtract the balance from the user.

## Withdrawing on L1

> After the transaction is completed on L2, the portal must call the outbox to successfully transfer funds to the user on L1. Like with deposits, things can be complex here where what happens if the transaction was done on L2 to burn tokens but can’t be withdrawn to L1. Then the funds are lost forever.

Reviewer: ```suggestion
After the transaction is completed on L2, the portal must call the outbox to successfully transfer funds to the user on L1. Like with deposits, things can be complex here. For example, what happens if the transaction was done on L2 to burn tokens but can’t be withdrawn to L1? Then the funds are lost forever! How do we prevent this?
```

Paste this in your `TokenPortal.sol`:

#include_code token_portal_withdraw /l1-contracts/test/portals/TokenPortal.sol solidity

Here we reconstruct the L2 to L1 message and check that this message exists on the outbox. If so, we consume it and transfer the funds to the recipient. As part of the reconstruction, the content hash looks similar to what we did in our bridge contract on aztec where we pass the amount and recipient to the the hash. This way a malicious actor can’t change the recipient parameter to the address and withdraw funds to themselves.

We also use a `_withCaller` parameter to determine the appropriate party that can execute this function on behalf of the recipient. If `withCaller` is false, then anyone can call the method and hence we use address(0), otherwise only msg.sender should be able to execute. This address should match the `callerOnL1` address we passed in aztec when withdrawing from L2.

> We call this pattern _designed caller_ which enables a new paradigm **where we can construct other such portals that talk to the token portal and therefore create more seamless crosschain legos** between L1 and L2.

Reviewer: Is the pattern called "designed" caller or "designated caller"? I guess either could make sense, just want to make sure it's correct.

Reviewer 2: the latter! "designated caller"
